File: WindowsCE/NurSample_vs2008/NurSample/Log.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Threading;
using System.Windows.Forms;
using System.IO;

using NurApiDotNet; //NurApi wrapper

namespace NurSample
{
    public partial class Log : UserControl
    {
        /// <summary>
        /// The NurApi handle
        /// </summary>
        NurApi hNur = null;

        /// <summary>
        /// Initializes a new instance of the <see cref="Inventory" /> class.
        /// </summary>
        public Log()
        {
            InitializeComponent();
            this.Enabled = false;
        }

        /// <summary> 
        /// Clean up any resources being used.
        /// </summary>
        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        /// <summary>
        /// Sets the NurApi.
        /// </summary>
        /// <param name="hNur">The handle of NurApi.</param>
        public void SetNurApi(NurApi hNur)
        {
            try
            {
                this.hNur = hNur;
                // Set event handlers for NurApi
                hNur.LogEvent += new EventHandler<NurApi.LogEventArgs>(hNur_LogEvent);
                // Update CheckBoxes
                int mask = hNur.GetLogLevel();
                logVerboseCheckBox.Checked = (mask & NurApi.LOG_VERBOSE) != 0;
                logErrorCheckBox.Checked = (mask & NurApi.LOG_ERROR) != 0;
                logUserCheckBox.Checked = (mask & NurApi.LOG_USER) != 0;
                logDataCheckBox.Checked = (mask & NurApi.LOG_DATA) != 0;
                // Enable controls
                this.Enabled = true;
            }
            catch (NurApiException ex)
            {
                MessageBox.Show(ex.ToString(), Program.appName);
            }
        }

        void hNur_LogEvent(object sender, NurApi.LogEventArgs e)
        {
            NurApi hNur = sender as NurApi;
            logListBox.Items.Add(string.Format("{0,8}: {1}", e.timestamp, e.message));
            while (logListBox.Items.Count > 500)
            {
                logListBox.Items.RemoveAt(0);
            }
            //logListBox.SelectedIndex = logListBox.Items.Count - 1;
        }

        private void logVerboseCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            int mask = hNur.GetLogLevel();
            if (logVerboseCheckBox.Checked)
            {
                mask |= NurApi.LOG_VERBOSE;
            }
            else
            {
                mask &= ~NurApi.LOG_VERBOSE;
            }
            hNur.SetLogLevel(mask);
        }

        private void logErrorCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            int mask = hNur.GetLogLevel();
            if (logErrorCheckBox.Checked)
            {
                mask |= NurApi.LOG_ERROR;
            }
            else
            {
                mask &= ~NurApi.LOG_ERROR;
            }
            hNur.SetLogLevel(mask);
        }

        private void logUserCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            int mask = hNur.GetLogLevel();
            if (logUserCheckBox.Checked)
            {
                mask |= NurApi.LOG_USER;
            }
            else
            {
                mask &= ~NurApi.LOG_USER;
            }
            hNur.SetLogLevel(mask);
        }

        private void logDataCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            int mask = hNur.GetLogLevel();
            if (logDataCheckBox.Checked)
            {
                mask |= NurApi.LOG_DATA;
            }
            else
            {
                mask &= ~NurApi.LOG_DATA;
            }
            hNur.SetLogLevel(mask);
        }

        private void clearButton_Click(object sender, EventArgs e)
        {
            logListBox.Items.Clear();
        }

        private void saveButton_Click(object sender, EventArgs e)
        {
            if (saveFileDialog1.ShowDialog() == DialogResult.OK)
            {
                using (StreamWriter sw = new StreamWriter(saveFileDialog1.FileName))
                {
                    for (int r = 0; r < logListBox.Items.Count; r++)
                    {
                        sw.WriteLine(logListBox.Items[r].ToString());
                    }
                    sw.Flush();
                }
            }
        }

        private void logToFileButton_Click(object sender, EventArgs e)
        {
            if (hNur.GetLogToFile() == false)
            {
                if (saveFileDialog1.ShowDialog() == DialogResult.OK)
                {
                    hNur.SetLogFilePath(saveFileDialog1.FileName);
                    hNur.SetLogToFile(true);
                    logToFileButton.Text = "Stop logging into file";
                }
            }
            else
            {
                hNur.SetLogToFile(false);
                logToFileButton.Text = "Log to file";
            }
        }
    }
}

File: WindowsCE/NurSample_vs2008_VB/NurSampleCE_VB/Log.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Text
Imports System.Threading
Imports System.Windows.Forms
Imports System.IO

Imports NurApiDotNet
'NurApi wrapper
Partial Public Class Log
    Inherits UserControl
    ''' <summary>
    ''' The NurApi handle
    ''' </summary>
    Private hNur As NurApi = Nothing

    ''' <summary>
    ''' Initializes a new instance of the <see cref="Inventory" /> class.
    ''' </summary>
    Public Sub New()
        InitializeComponent()
        Me.Enabled = False
    End Sub

    ''' <summary> 
    ''' Clean up any resources being used.
    ''' </summary>
    ''' <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
    Protected Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing AndAlso (components IsNot Nothing) Then
            components.Dispose()
        End If
        MyBase.Dispose(disposing)
    End Sub

    ''' <summary>
    ''' Sets the NurApi.
    ''' </summary>
    ''' <param name="hNur">The handle of NurApi.</param>
    Public Sub SetNurApi(ByVal hNur As NurApi)
        Try
            Me.hNur = hNur
            ' Set event handlers for NurApi
            AddHandler hNur.LogEvent, New EventHandler(Of NurApi.LogEventArgs)(AddressOf hNur_LogEvent)
            ' Update CheckBoxes
            Dim mask As Integer = hNur.GetLogLevel()
            logVerboseCheckBox.Checked = (mask And NurApi.LOG_VERBOSE) <> 0
            logErrorCheckBox.Checked = (mask And NurApi.LOG_ERROR) <> 0
            logUserCheckBox.Checked = (mask And NurApi.LOG_USER) <> 0
            logDataCheckBox.Checked = (mask And NurApi.LOG_DATA) <> 0
            ' Enable controls
            Me.Enabled = True
        Catch ex As NurApiException
            MessageBox.Show(ex.ToString(), Program.appName)
        End Try
    End Sub

    Private Sub hNur_LogEvent(ByVal sender As Object, ByVal e As NurApi.LogEventArgs)
        Dim hNur As NurApi = TryCast(sender, NurApi)
        logListBox.Items.Add(String.Format("{0,8}: {1}", e.timestamp, e.message))
        While logListBox.Items.Count > 500
            logListBox.Items.RemoveAt(0)
        End While
        'logListBox.SelectedIndex = logListBox.Items.Count - 1;
    End Sub

    Private Sub logVerboseCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
        Dim mask As Integer = hNur.GetLogLevel()
        If logVerboseCheckBox.Checked Then
            mask = mask Or NurApi.LOG_VERBOSE
        Else
            mask = mask And Not NurApi.LOG_VERBOSE
        End If
        hNur.SetLogLevel(mask)
    End Sub

    Private Sub logErrorCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
        Dim mask As Integer = hNur.GetLogLevel()
        If logErrorCheckBox.Checked Then
            mask = mask Or NurApi.LOG_ERROR
        Else
            mask = mask And Not NurApi.LOG_ERROR
        End If
        hNur.SetLogLevel(mask)
    End Sub

    Private Sub logUserCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
        Dim mask As Integer = hNur.GetLogLevel()
        If logUserCheckBox.Checked Then
            mask = mask Or NurApi.LOG_USER
        Else
            mask = mask And Not NurApi.LOG_USER
        End If
        hNur.SetLogLevel(mask)
    End Sub

    Private Sub logDataCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
        Dim mask As Integer = hNur.GetLogLevel()
        If logDataCheckBox.Checked Then
            mask = mask Or NurApi.LOG_DATA
        Else
            mask = mask And Not NurApi.LOG_DATA
        End If
        hNur.SetLogLevel(mask)
    End Sub

    Private Sub clearButton_Click(ByVal sender As Object, ByVal e As EventArgs)
        logListBox.Items.Clear()
    End Sub

    Private Sub saveButton_Click(ByVal sender As Object, ByVal e As EventArgs)
        If saveFileDialog1.ShowDialog() = DialogResult.OK Then
            Using sw As New StreamWriter(saveFileDialog1.FileName)
                For r As Integer = 0 To logListBox.Items.Count - 1
                    sw.WriteLine(logListBox.Items(r).ToString())
                Next
                sw.Flush()
            End Using
        End If
    End Sub

    Private Sub logToFileButton_Click(ByVal sender As Object, ByVal e As EventArgs)
        If hNur.GetLogToFile() = False Then
            If saveFileDialog1.ShowDialog() = DialogResult.OK Then
                hNur.SetLogFilePath(saveFileDialog1.FileName)
                hNur.SetLogToFile(True)
                logToFileButton.Text = "Stop logging into file"
            End If
        Else
            hNur.SetLogToFile(False)
            logToFileButton.Text = "Log to file"
        End If
    End Sub
End Class
